File: api/AltV.Net/Elements/Entities/VoiceChannel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using AltV.Net.CApi;
using AltV.Net.Elements.Pools;

namespace AltV.Net.Elements.Entities
{
    public class VoiceChannel : BaseObject, IVoiceChannel
    {
        public IntPtr VoiceChannelNativePointer { get; }

        private static IntPtr GetBaseObjectPointer(ICore core, IntPtr nativePointer)
        {
            unsafe
            {
                return core.Library.Server.VoiceChannel_GetBaseObject(nativePointer);
            }
        }

        public static uint GetId(IntPtr pedPointer)
        {
            unsafe
            {
                return Alt.Core.Library.Shared.VoiceChannel_GetID(pedPointer);
            }
        }

        public void AddPlayer(IPlayer player)
        {
            unsafe
            {
                CheckIfEntityExists();
                Core.Library.Server.VoiceChannel_AddPlayer(VoiceChannelNativePointer, player.PlayerNativePointer);
            }
        }

        public void RemovePlayer(IPlayer player)
        {
            unsafe
            {
                CheckIfEntityExists();
                Core.Library.Server.VoiceChannel_RemovePlayer(VoiceChannelNativePointer, player.PlayerNativePointer);
            }
        }

        public void MutePlayer(IPlayer player)
        {
            unsafe
            {
                CheckIfEntityExists();
                Core.Library.Server.VoiceChannel_MutePlayer(VoiceChannelNativePointer, player.PlayerNativePointer);
            }
        }

        public void UnmutePlayer(IPlayer player)
        {
            unsafe
            {
                CheckIfEntityExists();
                Core.Library.Server.VoiceChannel_UnmutePlayer(VoiceChannelNativePointer, player.PlayerNativePointer);
            }
        }

        public bool HasPlayer(IPlayer player)
        {
            unsafe
            {
                CheckIfEntityExists();
                return Core.Library.Server.VoiceChannel_HasPlayer(VoiceChannelNativePointer, player.PlayerNativePointer) == 1;
            }
        }

        public bool IsPlayerMuted(IPlayer player)
        {
            unsafe
            {
                CheckIfEntityExists();
                return Core.Library.Server.VoiceChannel_IsPlayerMuted(VoiceChannelNativePointer, player.PlayerNativePointer) == 1;
            }
        }

        public bool IsSpatial
        {
            get
            {
                unsafe
                {
                    CheckIfEntityExists();
                    return Core.Library.Server.VoiceChannel_IsSpatial(VoiceChannelNativePointer) == 1;
                }
            }
        }

        public float MaxDistance
        {
            get
            {
                unsafe
                {
                    CheckIfEntityExists();
                    return Core.Library.Server.VoiceChannel_GetMaxDistance(VoiceChannelNativePointer);
                }
            }
        }

        public uint Filter
        {
            get
            {
                unsafe
                {
                    CheckIfEntityExistsOrCached();
                    return Core.Library.Server.VoiceChannel_GetFilter(VoiceChannelNativePointer);
                }
            }
            set
            {
                unsafe
                {
                    CheckIfEntityExists();
                    Core.Library.Server.VoiceChannel_SetFilter(VoiceChannelNativePointer, value);
                }
            }
        }

        public int Priority
        {
            get
            {
                unsafe
                {
                    CheckIfEntityExistsOrCached();
                    return Core.Library.Server.VoiceChannel_GetPriority(VoiceChannelNativePointer);
                }
            }
            set
            {
                unsafe
                {
                    CheckIfEntityExists();
                    Core.Library.Server.VoiceChannel_SetPriority(VoiceChannelNativePointer, value);
                }
            }
        }

        public IReadOnlyCollection<IPlayer> Players
        {
            get
            {
                unsafe
                {
                    CheckIfCallIsValid();
                    ulong size = 0;
                    var ptr = Core.Library.Server.VoiceChannel_GetPlayers(VoiceChannelNativePointer, &size);
                    var data = new IntPtr[size];
                    Marshal.Copy(ptr, data, 0, (int)size);
                    var arr = data.Select(e => (IPlayer)Core.PoolManager.GetOrCreate(Core, e, BaseObjectType.Player))
                        .ToArray();
                    Core.Library.Shared.FreePlayerArray(ptr);
                    return arr;
                }
            }
        }

        public ulong PlayerCount
        {
            get
            {
                unsafe
                {
                    CheckIfEntityExistsOrCached();
                    return Core.Library.Server.VoiceChannel_GetPlayerCount(VoiceChannelNativePointer);
                }
            }
        }

        public VoiceChannel(ICore core, IntPtr nativePointer, uint id) : base(core, GetBaseObjectPointer(core, nativePointer), BaseObjectType.VoiceChannel, id)
        {
            VoiceChannelNativePointer = nativePointer;
        }
    }
}

File: runtime/src/altv-c-api/voice_channel.cpp
#include "voice_channel.h"

alt::MValueConst* VoiceChannel_GetMetaData(alt::IVoiceChannel* voiceChannel, const char* key) {
    return new alt::MValueConst(voiceChannel->GetMetaData(alt::String(key)));
}

void VoiceChannel_SetMetaData(alt::IVoiceChannel* channel, const char* key, alt::MValueConst* val) {
    if (val == nullptr) return;
    channel->SetMetaData(alt::String(key), val->Get()->Clone());
}

uint8_t VoiceChannel_HasMetaData(alt::IVoiceChannel* voiceChannel, const char* key) {
    return voiceChannel->HasMetaData(alt::String(key));
}

void VoiceChannel_DeleteMetaData(alt::IVoiceChannel* voiceChannel, const char* key) {
    voiceChannel->DeleteMetaData(alt::String(key));
}

void VoiceChannel_AddRef(alt::IVoiceChannel* voiceChannel) {
    voiceChannel->AddRef();
}

void VoiceChannel_RemoveRef(alt::IVoiceChannel* voiceChannel) {
    voiceChannel->RemoveRef();
}

void VoiceChannel_AddPlayer(alt::IVoiceChannel* channel, alt::IPlayer* player) {
    channel->AddPlayer(player);
}

void VoiceChannel_RemovePlayer(alt::IVoiceChannel* channel, alt::IPlayer* player) {
    channel->RemovePlayer(player);
}

void VoiceChannel_MutePlayer(alt::IVoiceChannel* channel, alt::IPlayer* player) {
    channel->MutePlayer(player);
}

void VoiceChannel_UnmutePlayer(alt::IVoiceChannel* channel, alt::IPlayer* player) {
    channel->UnmutePlayer(player);
}

uint8_t VoiceChannel_IsPlayerConnected(alt::IVoiceChannel* channel, alt::IPlayer* player) {
    return channel->HasPlayer(player);
}

uint8_t VoiceChannel_HasPlayer(alt::IVoiceChannel* channel, alt::IPlayer* player) {
    return channel->HasPlayer(player);
}

uint8_t VoiceChannel_IsPlayerMuted(alt::IVoiceChannel* channel, alt::IPlayer* player) {
    return channel->IsPlayerMuted(player);
}

uint8_t VoiceChannel_IsSpatial(alt::IVoiceChannel* channel) {
    return channel->IsSpatial();
}

float VoiceChannel_GetMaxDistance(alt::IVoiceChannel* channel) {
    return channel->GetMaxDistance();
}
